src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc

```diff
@@ -317,270 +260,228 @@ static void pf_update_locked(grpc_exec_ctx *exec_ctx, grpc_lb_policy *policy,
-      const grpc_lb_policy_args *args = p->pending_update_args;
-      p->pending_update_args = NULL;
-      pf_update_locked(exec_ctx, &p->base, args);
-    }
+  // If the policy is shutting down, unref and return.
+  if (p->shutdown) {
+    grpc_lb_subchannel_data_stop_connectivity_watch(exec_ctx, sd);
+    grpc_lb_subchannel_data_unref_subchannel(exec_ctx, sd, "pf_shutdown");
+    grpc_lb_subchannel_list_unref_for_connectivity_watch(
+        exec_ctx, sd->subchannel_list, "pf_shutdown");
     return;
   }
-  GRPC_ERROR_REF(error);
-  if (p->shutdown) {
-    GRPC_LB_POLICY_WEAK_UNREF(exec_ctx, &p->base, "pick_first_connectivity");
-    GRPC_ERROR_UNREF(error);
+  // If the subchannel list is shutting down, stop watching.
+  if (sd->subchannel_list->shutting_down || error == GRPC_ERROR_CANCELLED) {
+    grpc_lb_subchannel_data_stop_connectivity_watch(exec_ctx, sd);
+    grpc_lb_subchannel_data_unref_subchannel(exec_ctx, sd, "pf_sl_shutdown");
+    grpc_lb_subchannel_list_unref_for_connectivity_watch(
+        exec_ctx, sd->subchannel_list, "pf_sl_shutdown");
     return;
-  } else if (p->selected != NULL) {
-    if (p->checking_connectivity == GRPC_CHANNEL_TRANSIENT_FAILURE) {
-      /* if the selected channel goes bad, we're done */
-      p->checking_connectivity = GRPC_CHANNEL_SHUTDOWN;
-    }
-    grpc_connectivity_state_set(exec_ctx, &p->state_tracker,
-                                p->checking_connectivity, GRPC_ERROR_REF(error),
-                                "selected_changed");
-    if (p->checking_connectivity != GRPC_CHANNEL_SHUTDOWN) {
-      grpc_connected_subchannel_notify_on_state_change(
-          exec_ctx, p->selected, p->base.interested_parties,
-          &p->checking_connectivity, &p->connectivity_changed);
+  }
+  // If we're still here, the notification must be for a subchannel in
+  // either the current or latest pending subchannel lists.
+  GPR_ASSERT(sd->subchannel_list == p->subchannel_list ||
+             sd->subchannel_list == p->latest_pending_subchannel_list);
+  // Update state counters.
+  sd->curr_connectivity_state = sd->pending_connectivity_state_unsafe;
+  // Handle updates for the currently selected subchannel.
+  if (p->selected == sd) {
+    // If the new state is anything other than READY and there is a
+    // pending update, switch to the pending update.
+    if (sd->curr_connectivity_state != GRPC_CHANNEL_READY &&
+        p->latest_pending_subchannel_list != NULL) {
+      p->selected = NULL;
+      grpc_lb_subchannel_list_shutdown_and_unref(
+          exec_ctx, p->subchannel_list, "selected_not_ready+switch_to_update");
+      p->subchannel_list = p->latest_pending_subchannel_list;
+      p->latest_pending_subchannel_list = NULL;
+      grpc_lb_subchannel_data *new_sd =
+          &p->subchannel_list
+               ->subchannels[p->subchannel_list->checking_subchannel];
+      grpc_connectivity_state_set(
+          exec_ctx, &p->state_tracker, new_sd->curr_connectivity_state,
```

Comment:
Whoops, I somehow missed this comment in my last pass.

We only execute this code if `p->selected == sd`, meaning that we've already chosen a subchannel (because it went into READY state) and this is an update for that subchannel.  In the current code, when we choose a subchannel, we unref all of the other subchannels in the subchannel list, thus committing to using only the selected subchannel, at least until we get a new update.  Thus, if the selected subchannel goes into state SHUTDOWN, we have no choice but to put ourselves in that state.

There are two future changes that will improve this.  One of them is your PR to have the LB policy request re-resolution without putting itself in state SHUTDOWN, and I would expect your PR to modify the code in this location (indeed, I think it's already doing the right thing in this case, although you'll have to retain that when you merge the changes from this PR).  The other future change that I would like to make is to change pick_first to keep the other subchannels around so that it can try some of the other addresses if the originally selected address fails and we have not yet received an update.  But that will be done in a future PR.

For this PR, I am leaving the existing logic here essentially the same.

Reply:
Good catch!  I've changed it to always put the LB policy into TRANSIENT_FAILURE in this case.